core/src/win32/plugins/filed/mssqlvdi-fd.cc

```diff
@@ -1475,9 +1476,25 @@ static inline bool TearDownVdiDevice(PluginContext* ctx, io_pkt* io)
   // Check if the VDI device is closed.
   if (p_ctx->VDIDevice) {
     hr = p_ctx->VDIDevice->GetCommand(VDI_WAIT_TIMEOUT, &cmd);
+    // Return Value	Explanation
+    // NOERROR	A command was fetched.
+    // VD_E_CLOSE	The device has been closed by the server.
+    // VD_E_TIMEOUT	No command was available and the time-out expired.
+    // VD_E_ABORT	Either the client or the server has used the SignalAbort
+    // to force a shutdown.
+    static std::map<int, std::string> VdiRetVals{
+        {VD_E_CLOSE, "VD_E_CLOSE"},
+        {VD_E_TIMEOUT, "VD_E_TIMEOUT"},
+        {VD_E_ABORT, "VD_E_ABORT"},
+        {NO_ERROR, "NO_ERROR"},
+    };
     if (hr != VD_E_CLOSE) {
-      Jmsg(ctx, M_ERROR, "Abnormal termination, VDIDevice not closed.\n");
-      Dmsg(ctx, debuglevel, "Abnormal termination, VDIDevice not closed.\n");
+      Jmsg(ctx, M_ERROR,
+           "Abnormal termination, VDIDevice not closed. Result = %s\n",
+           VdiRetVals[hr].c_str());
+      Dmsg(ctx, debuglevel,
+           "Abnormal termination, VDIDevice not closed. Result = %s\n",
+           VdiRetVals[hr].c_str());
```

Comment:
While this is very easy to read, its also a race condition/crash waiting to happen. `VdiRetVals[hr]` _inserts_ the empty string into the map if `hr` is not found inside the map already. If any other job was inside this function as well then bad things will happen.

Something like this should be safer:
```
auto found = VdiRetVals.find(hr);
std::string error_string;
if (found != VdiRetVals.end()) {
  error_string = found->second.c_str();
} else {
  error_string = "Unknown error (res=";
  error_string += std::to_string(hr);
  error_string += ")";
}
```
This would also allow you to define the map as `const`.